### PR TITLE
feat: improve tab rename and drag interactions

### PR DESCRIPTION
**Pull request by sundy-li**
## Summary
- add inline rename for horizontal tabs plus sidebar blur-save, Escape-cancel, and focus-time select-all
- keep rename commits stable across tab reorder/close, seed rename inputs from rendered labels, and refocus the active terminal after successful renames
- improve horizontal tab drag-reorder with browser-style left/right slots, a real-bounds trailing drop target, and macOS-safe drag behavior

## Test Plan
- [x] cargo test -p con workspace::tests:: -- --nocapture
- [x] cargo build -p con
- [x] manually verify horizontal tab rename selects existing text on focus and blur-without-edits preserves rendered labels
- [x] manually verify sidebar rename saves on blur, Escape cancels, and successful renames refocus the active terminal
- [x] manually verify tab drag reorder works in both directions and dropping after the last tab uses the real trailing target without dragging the whole window

## Summary by CodeRabbit

* **New Features**
  * Added inline renaming for tabs with double-click to initiate edits
  * Enhanced sidebar renaming with improved validation and user label normalization
  * Improved tab drag-and-drop reordering with clearer drop indicators

* **Bug Fixes**
  * Fixed stale rename handlers from committing cancelled edits through generation tracking
  * Resolved focus issues by refocusing active terminal after successful renames

* **Documentation**
  * Added design specification for sidebar rename and terminal refocus behavior

### CODE REVIEW

File: crates/con-app/src/sidebar.rs [hunk 1]
```diff
@@ -158,6 +160,9 @@ pub struct SessionSidebar {
     effective_panel_max_width: f32,
     /// Inline rename state, `Some` while the user is editing a label.
     rename: Option<RenameState>,
+    /// Escape-cancel marker so a subsequent blur from the old input
+    /// does not commit the just-cancelled rename.
+    rename_cancelled_session_id: Option<u64>,
```
[the remark this answers is not in the document]
Comment (sundy-li) [reply crates/con-app/src/sidebar.rs]: Fixed in crates/con-app/src/workspace.rs.

The horizontal tab rename flow now captures the tab's stable summary_id instead of the original numeric index:
- TabRenameEditor carries tab_id plus render-time tab_index
- PressEnter / Blur commit through commit_tab_rename(tab_id, ...)
- Escape cancel now uses tab_rename_cancelled_id: Option<u64>
- close/reorder remap only updates render positioning, not identity

That addresses the real bug here: stale rename events after reorder/close can no longer commit against the wrong tab slot.

I did not apply the sidebar generation-token suggestion from this comment. For the sidebar path, the current code already commits by stable session_id rather than list index, so the specific reorder-invalidates-index issue called out in this thread does not apply there. This PR only resolves the horizontal tab-strip issue the thread was filed on.

[the remark this answers is not in the document]
Comment (sundy-li) [reply crates/con-app/src/sidebar.rs]: Fixed the workspace-side issue in crates/con-app/src/workspace.rs.

The horizontal tab-strip rename editor is now seeded from the rendered label path instead of user_label || raw terminal title. That means focus→blur with no edits preserves AI/SSH/CWD-derived naming instead of materializing a new explicit label from the raw title.

Concretely, begin_tab_rename() now derives the initial value through the same smart_tab_presentation() inputs the strip uses, via a shared tab_rename_initial_label() helper.

Verification: cargo test -p con workspace::tests:: -- --nocapture (8 passed, 0 failed).

[the remark this answers is not in the document]
Comment (sundy-li) [reply crates/con-app/src/sidebar.rs]: Both workspace findings are covered in the current code.

1) Rendered label seeding
- begin_tab_rename() no longer seeds from user_label || raw terminal title
- it now derives the initial value through the same smart_tab_presentation() inputs the strip uses, via tab_rename_initial_label()
- this preserves AI/SSH/CWD-derived labels on focus→blur with no edits

2) Stable-id commit
- the InputEvent subscriber no longer closes over the original index
- it captures tab_id = tab.summary_id and commits through commit_tab_rename(tab_id, ...)
- commit_tab_rename() resolves the current slot with tab_index_for_summary_id(tab_id) before mutating, and the close/reorder remap only updates render-time tab_index

Verification from the current branch:
- cargo test -p con workspace::tests:: -- --nocapture
- result: 8 passed, 0 failed

I am resolving this thread as addressed for the workspace path. The separate sidebar Escape-then-reopen race is a different issue and is not one of the two workspace findings called out here.

[the remark this answers is not in the document]
Comment (sundy-li) [reply crates/con-app/src/sidebar.rs]: Verified both findings against the current branch.

1) Horizontal tab trailing drop slot: still valid, fixed with a minimal change in crates/con-app/src/workspace.rs. The container-level fallback no longer uses the fixed approx_tab_w estimate. Instead, the after-last-tab slot is now handled by a dedicated trailing drop target element that uses its real bounds, and the container fallback only clears a stale trailing slot when the pointer moves back inside the strip. Validation: cargo test -p con workspace::tests:: -- --nocapture (9 passed, 0 failed).

2) Doc spelling cleanup: not a remaining issue in current prose. The narrative text in docs/impl/sidebar-rename-blur-terminal-refocus-design-2026-05-04.md already uses US spelling (canceled). The remaining cancelled occurrences are code identifiers such as tab_rename_cancelled_index / rename_cancelled_session_id, and the review explicitly says to leave those tokens unchanged. No doc edit was needed.

**Comment by wey-gu** [comment]
@codex review

File: crates/con-app/src/sidebar.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sundy-li) [reply crates/con-app/src/sidebar.rs]: Fixed the remaining sidebar race in crates/con-app/src/sidebar.rs.

The sidebar rename flow no longer keys Escape-cancel by session_id alone. RenameState now carries a monotonic generation, SessionSidebar tracks rename_generation plus rename_cancelled_generation, and PressEnter / Blur only commit if the firing input still matches the active generation. That closes the Escape → immediate reopen → stale Blur window on the same tab.

Validation from the current branch:
- cargo test -p con sidebar::tests:: -- --nocapture
- cargo test -p con workspace::tests:: -- --nocapture
- result: sidebar 2 passed, workspace 9 passed, 0 failed

For the doc spelling note: current prose already uses US spelling (canceled); the remaining cancelled occurrences are intentional code identifiers, which the review explicitly asked us not to rename.

[the remark this answers is not in the document]
Comment (sundy-li) [reply crates/con-app/src/sidebar.rs]: Fixed the remaining sidebar race in crates/con-app/src/sidebar.rs.

The issue was valid: the old guard was keyed only by session_id, so Escape on tab X followed by immediately reopening rename on the same tab X could clear the guard before the stale input's delayed Blur arrived.

The sidebar rename flow now uses a per-editor generation token:
- RenameState carries generation
- SessionSidebar tracks rename_generation and rename_cancelled_generation
- PressEnter / Blur only commit if the firing input still matches the active rename generation
- Escape records the canceled generation, so a stale Blur from that editor is ignored even after same-tab reopen

Validation from the current branch:
- cargo test -p con sidebar::tests:: -- --nocapture
- cargo fmt --all && cargo test -p con sidebar::tests:: -- --nocapture
- result: 2 passed, 0 failed

**Comment by sundy-li** [comment]
@codex review

File: crates/con-app/src/workspace.rs [hunk 2]
```diff
@@ -4763,6 +4792,216 @@ impl ConWorkspace {
...
+            }
+        })
+        .detach();
+
+        self.tab_rename = Some(TabRenameEditor {
+            tab_id,
+            tab_index: index,
+            input: input.clone(),
+        });
+        self.tab_rename_cancelled_id = None;
```
[the remark this answers is not in the document]
Comment (sundy-li) [reply crates/con-app/src/workspace.rs]: Fixed in e62d33d. The horizontal tab rename flow now uses a per-editor generation token instead of a tab-id cancel marker: begin_tab_rename allocates a new generation, Escape records the canceled generation, and Blur/Enter only commit when the event still belongs to the active generation. That prevents a delayed blur from a just-canceled editor from tearing down or committing after an immediate reopen on the same tab. Validated with `cargo test --manifest-path "/Users/sundy/work/con-terminal/.worktrees/tab-rename/crates/con-app/Cargo.toml" --quiet` (64 passed).

**Comment by wey-gu** [comment]
@codex review

File: crates/con-app/src/workspace.rs [hunk 3]
```diff
@@ -10225,6 +10483,15 @@ impl Render for ConWorkspace {
...
                 .window_control_area(WindowControlArea::Drag)
+                .on_mouse_down(MouseButton::Left, |_, window, _cx| {
+                    // NSWindow.isMovable is set to false so that GPUI
+                    // element drags (tab reorder, etc.) don't also move
+                    // the window. We restore the drag-to-move gesture
+                    // here on the top_bar background; child elements
+                    // with on_mouse_down (tabs, buttons) consume the
+                    // event before it reaches this handler.
+                    window.start_window_move();
+                })
```
[the remark this answers is not in the document]
Comment (sundy-li) [reply crates/con-app/src/workspace.rs]: Fixed in 72af501. The root cause was that several top-bar controls (`tab-new`, `toggle-vertical-tabs`, `toggle-input-bar`, `toggle-agent-panel`, and non-macOS `toggle-settings`) only handled `on_click`, while the macOS titlebar container starts `start_window_move()` on left mouse-down. I kept the drag behavior on the titlebar background, but added explicit left-button `on_mouse_down` handlers on those controls that call `cx.stop_propagation()` before their existing click handlers run, so control clicks are no longer stolen by the parent drag handler. Validated with `cargo test --manifest-path "/Users/sundy/work/con-terminal/.worktrees/tab-rename/crates/con-app/Cargo.toml" --quiet` (65 passed), and added a regression test that asserts these top-bar controls consume left mouse-down in source.

**Comment by wey-gu** [comment]
Applied a follow-up review fix commit: `ca579ad3 fix: tighten tab rename interactions`.

What changed:
- Rename blur now remains the commit path when switching tabs; only Escape cancels.
- Focus→blur with no edits restores terminal focus without pinning smart AI/SSH/CWD tab labels as explicit user labels.
- Horizontal/sidebar rename commit paths now track actual input changes.
- macOS titlebar drag now follows Zed's delayed-drag pattern so double-click titlebar behavior is preserved while tab drag still works.
- Updated the design note and bumped CHANGELOG to `v0.1.0-beta.61`.

Validated locally:
- `cargo fmt --all`
- `cargo test -p con sidebar::tests:: -- --nocapture`
- `cargo test -p con workspace::tests:: -- --nocapture`
- `cargo test -p con --quiet`
- `cargo check -p con`

**Comment by wey-gu** [comment]
@codex review

**Comment by wey-gu** [comment]
THNKS for the amazing work 🙇
